eth/filters/filter_test.go

```diff
@@ -102,140 +107,264 @@ func BenchmarkFilters(b *testing.B) {
+			f: sys.NewRangeFilter(-1, -3, nil, nil),
 		}, {
-			sys.NewRangeFilter(-4, -4, nil, nil), nil,
+			f:   sys.NewRangeFilter(-4, -1, nil, nil),
+			err: "safe header not found",
 		}, {
-			sys.NewRangeFilter(-1, -4, nil, nil), nil,
+			f:   sys.NewRangeFilter(-4, -4, nil, nil),
+			err: "safe header not found",
+		}, {
+			f:   sys.NewRangeFilter(-1, -4, nil, nil),
+			err: "safe header not found",
+		}, {
+			f:    sys.NewRangeFilter(-2, -2, nil, nil),
+			want: `[{"address":"0xfe00000000000000000000000000000000000000","topics":["0x0000000000000000000000000000000000000000000000000000746f70696335"],"data":"0x","blockNumber":"0x3e9","transactionHash":"0x4110587c1b8d86edc85dce929a34127f1cb8809515a9f177c91c866de3eb0638","transactionIndex":"0x0","blockHash":"0xc7245899e5817f16fa99cf5ad2d9c1e4b98443a565a673ec9c764640443ef037","logIndex":"0x0","removed":false}]`,
+		}, {
+			f:    sys.NewRangeFilter(-1, -2, nil, nil),
+			want: `[{"address":"0xfe00000000000000000000000000000000000000","topics":["0x0000000000000000000000000000000000000000000000000000746f70696334"],"data":"0x","blockNumber":"0x3e8","transactionHash":"0x9a87842100a638dfa5da8842b4beda691d2fd77b0c84b57f24ecfa9fb208f747","transactionIndex":"0x0","blockHash":"0xb360bad5265261c075ece02d3bf0e39498a6a76310482cdfd90588748e6c5ee0","logIndex":"0x0","removed":false},{"address":"0xfe00000000000000000000000000000000000000","topics":["0x0000000000000000000000000000000000000000000000000000746f70696335"],"data":"0x","blockNumber":"0x3e9","transactionHash":"0x4110587c1b8d86edc85dce929a34127f1cb8809515a9f177c91c866de3eb0638","transactionIndex":"0x0","blockHash":"0xc7245899e5817f16fa99cf5ad2d9c1e4b98443a565a673ec9c764640443ef037","logIndex":"0x0","removed":false}]`,
+		}, {
+			f:   sys.NewRangeFilter(-2, -1, nil, nil),
+			err: "invalid block range",
 		},
 	} {
-		logs, _ := tc.f.Logs(context.Background())
-		var haveHashes []common.Hash
-		for _, l := range logs {
-			haveHashes = append(haveHashes, l.Topics[0])
-		}
-		if have, want := len(haveHashes), len(tc.wantHashes); have != want {
-			t.Fatalf("test %d, have %d logs, want %d", i, have, want)
+		logs, err := tc.f.Logs(context.Background())
+		if err == nil && tc.err != "" {
+			t.Fatalf("test %d, expected error %q, got nil", i, tc.err)
+		} else if err != nil && err.Error() != tc.err {
+			t.Fatalf("test %d, expected error %q, got %q", i, tc.err, err.Error())
 		}
-		if len(haveHashes) == 0 {
+		if tc.want == "" && len(logs) == 0 {
 			continue
 		}
-		if !reflect.DeepEqual(tc.wantHashes, haveHashes) {
-			t.Fatalf("test %d, have %v want %v", i, haveHashes, tc.wantHashes)
+		have, err := json.Marshal(logs)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if string(have) != tc.want {
+			t.Fatalf("test %d, have:\n%s\nwant:\n%s", i, have, tc.want)
 		}
 	}
+
+	t.Run("timeout", func(t *testing.T) {
+		f := sys.NewRangeFilter(0, -1, nil, nil)
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Microsecond)
```

Comment:
```suggestion
		ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(-time.Hour))
```
Even less flaky, because if the time is already passed, the cancel is invoked immediately, (`context.go`:)
```
	if dur <= 0 {
		c.cancel(true, DeadlineExceeded, nil) // deadline has already passed
		return c, func() { c.cancel(false, Canceled, nil) }
```